plugin/firewall/README.md

```diff
@@ -0,0 +1,112 @@
+
+* **EXPRESSION** defines the expression to evaluate in order to validate the action and interrupt the sequence of rules.
+Expression is a [go-like language](https://github.com/Knetic/govaluate/blob/master/MANUAL.md)
+where the variables are either the `metadata` of CoreDNS either a list of names associated with the DNS query/response information.
+Usual operators applies.
+
+Exemples of expression using the usual metadata:
+* `client_ip == '10.0.0.20'`
+* `type == 'AAAA'`
+* `type IN ('AAAA', 'A', 'TXT')`
+* `type IN ('AAAA', 'A') && name =~ 'google.com'`
+* `[mac/address] =~ '.*:FF:.*'`
+
+NOTE: because of the `/` separator included in a label of metadata, those labels must be enclosed on bracket [...] for a correct evaluation by the expression engine
+
+The following names are supported for querying information
+
+* `type`: type of the request (A, AAAA, TXT, ..)
+* `name`: name of the request (the domain requested)
+* `class`: class of the request (IN, CS, CH, ...)
+* `proto`: protocol used (tcp or udp)
+* `remote`: client's IP address, for IPv6 addresses these are enclosed in brackets: `[::1]`
+* `size`: request size in bytes
+* `port`: client's port
+* `duration`: response duration
+* `rcode`: response CODE (NOERROR, NXDOMAIN, SERVFAIL, ...)
+* `rsize`: raw (uncompressed), response size (a client may receive a smaller response)
+* `>rflags`: response flags, each set flag will be displayed, e.g. "aa, tc". This includes the qr
+  bit as well
+* `>bufsize`: the EDNS0 buffer size advertised in the query
+* `>do`: is the EDNS0 DO (DNSSEC OK) bit set in the query
+* `>id`: query ID
+* `>opcode`: query OPCODE
+* `server_ip`: server's IP address, for IPv6 addresses these are enclosed in brackets: `[::1]`
+* `server_port` : client's port
+* `response_ip` : the IP returned in the first A or AAAA record of the Answer section
+
+
+## Examples
+
+Allow queries for example.com.
+Allow also the queries for google.com if those are A or AAAA type of queries
+NXDOMAIN every other queries
+
+~~~ corefile
+. {
+   firewall query {
+      allow name =~ 'example.com'
+      allow name =~ 'google.com' && (type == 'A' || type == 'AAAA')
+      block true
+   }
+}
+~~~
+
+
+Define the metadata labels `group_id` and `client_id` with values extracted from EDNS0.
+and use those values to filter the DNS queries: any query that doesn't have the group_id of `123456789` AND doesn't have the client_id of `ABCDEF` will be returned REFUSED.
+In addition, respond with a REFUSE if the first `A`/`AAAA` record in the response contains an IP in 172.217.0.0/16.
+
+
+~~~ txt
+example.org {
+   metadata {
+      group_id edns0 0xffed bytes
+      client_id edns0 0xffee bytes
+   }
+   firewall query {
+      refuse [metadata/client_id] != 'ABCDEF'
+      refuse [metadata/group_id] != '123456789'
+      block true
+   }
+   firewall response {
+      refuse  response_ip =~ '172.217.*'   # refuse any IP that is in 172.217.0.0/16
+      allow true
+   }
+}
+~~~
```

Comment:
Add `See Also` section to point to the language.

What does a minimal firewall plugin look like? I.e a block all example and a allow allow example would help here.

(the current file also has a spurious empty line on the end)